gobblin-utility/src/main/java/org/apache/gobblin/util/filesystem/DataFileVersion.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.util.filesystem;
+
+import com.typesafe.config.Config;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Set;
+
+import org.apache.gobblin.util.ClassAliasResolver;
+import org.apache.gobblin.util.ConfigUtils;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+
+/**
+ * An interface to set and get "versions" to data files.
+ *
+ * The version is a rough signature to the data contents. It allows data preserving functionality (like copy) to replicate
+ * the version independently of metadata with other semantics like file modification time.
+ *
+ * Examples where this might be useful is data syncing between two locations. Relying on modification times to detect
+ * data changes may lead to a feedback loop of copying: data gets created at location A at time 0,
+ * at time 1 data is copied to location B, sync mechanism might incorrectly believe that since mod time of location B
+ * is higher, it should be synced back to location A, etc.
+ *
+ * Required properties:
+ * - REPLICABLE: Two calls to `getVersion` on a file that has clearly not been modified must return the same version.
+ * - MONOTONOUS: The default version of a file is an increasing function of modification time.
+ * - CONSERVATIVE: If file f had its version last set to v, but the versioning implementation determines the file MIGHT
+ *                 have been modified and it chooses to return a version, it will return a value strictly larger than v.
+ *
+ * A common pattern to achieve monotonicity and conservativeness will be to invalidate the version of a data file
```

Comment:
What does 'invalidate' imply here? Does that infer the getVersion will return a new version? In this case if we have STRICT will it return exception?

Reply:
It means to fallback to default version if anything external has modified the file I guess? I can rephrase this.

STRICT does mean that if a modification was made, `getVersion` will return an error to indicate that.